### PR TITLE
Sample: event jsvm

### PR DESCRIPTION
**Pull request by slaff**
This sample demonstrates how to run your JavaScript application as performing some action
in response to an event, input or stimulus. 
Since Sming framework itself has an [event-driven architecture](https://sming.readthedocs.io/en/latest/information/events.html) such JavaScript applications will be able to use all the advantages that Sming has in terms of CPU, RAM and power usage.

### CODE REVIEW

**Comment by slaff** [comment]
@mikee47 Can you review also this PR when you have time?

File: README.rst [hunk 1]
```diff
@@ -26,14 +26,16 @@ Samples
 -------
 
 - :sample:`Basic_Sample` - demonstrates how to use JerryScript and run compiled JavaScript code on a micro-controller.
-- :sample:`Advanced_Sample` - demonstrates how to modify and compile JavaScript code on the fly using only your browser and run that code on a micro-controller.  
+- :sample:`Advanced_Sample` - demonstrates how to modify and compile JavaScript code on the fly using only your browser and run that code on a micro-controller.
+- :sample:`Event_Sample` - demonstrates how to write JavaScript applications using event-driven architecture.
```
Comment (mikee47) [inline README.rst]: Suggest removing this section as page template includes a list of samples https://sming.readthedocs.io/en/latest/_inc/Sming/Libraries/jerryscript/index.html#used-by. Also note `:sample:` will only pick up framework samples - needs looking at separately.

File: component.mk [hunk 2]
```diff
@@ -12,7 +12,7 @@ COMPONENT_INCDIRS := \
 	$(JERRYSCRIPT_ROOT)/jerry-core \
 	$(COMPONENT_SRCDIRS)
 
-COMPONENT_DOCFILES := jerryscript/docs/img/engines_high_level_design.png
+COMPONENT_DOCFILES := jerryscript/docs/img/engines_high_level_design.png JerryScriptUpdate.rst
```
Comment (mikee47) [inline component.mk]: Revert - all .rst files in root directory are picked up automatically. https://sming.readthedocs.io/en/latest/information/develop/documentation.html#source-file-location

File: samples/Event_Jsvm/component.mk [hunk 3]
```diff
@@ -0,0 +1,11 @@
+COMPONENT_DEPENDS := jerryscript
+
+COMPONENT_SRCDIRS += src/
```
Comment (mikee47) [inline samples/Event_Jsvm/component.mk]: `COMPONENT_SRCDIRS := src` no /

File: samples/Event_Jsvm/files/main.js [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+
+	addEventListener("EVENT_TEMP", function(event) {
+		// You can register more than one listener to an event
+		print('Second listener');
+	});
+
+	// You can also run a normal function as callback instead of
+	// an anonymous function.
+	addEventListener("EVENT_TEMP", thirdListener);
+}
```
Comment (mikee47) [inline samples/Event_Jsvm/files/main.js]: CR missing

File: README.rst [hunk 5]
```diff
@@ -66,4 +68,3 @@ Credits
 -------
 
 The initial work on the JerryScript library for Sming was done as part of the `U:Kit project <https://github.com/attachix/ukit>`_. 
-
```
Comment (mikee47) [inline README.rst:69]: Missing CR

File: samples/Event_Jsvm/src/include/vm_functions.h [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+ * All files of the Sming Core are provided under the LGPL v3 license.
+ *
+ * jsvm-ext.cpp
+ *
+ * @author Nov 2021 - Slavey Karadzhov <slav@attachix.com>
+ */
+
+#pragma once
+
+#include <jerry-core/include/jerryscript.h>
```
Comment (mikee47) [inline samples/Event_Jsvm/src/include/vm_functions.h]: `#include <jerryscript.h>` is sufficient

File: samples/Event_Jsvm/src/include/vm_functions.h [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+#include <WString.h>
+
+#ifdef __cplusplus
+
+using JsEventData = HashMap<String, String>;
+
+void triggerEvent(const String& name, const JsEventData& data);
+
+extern "C" {
+#endif
```
Comment (mikee47) [inline samples/Event_Jsvm/src/include/vm_functions.h]: remove

File: samples/Event_Jsvm/src/include/vm_functions.h [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+ * 
+ * @endcode
+ * 
+ */
+jerry_value_t addEventListener(const jerry_value_t function_obj, const jerry_value_t this_val,
+							   const jerry_value_t args_p[], const jerry_length_t args_cnt);
+
+#ifdef __cplusplus
+}
+#endif
```
Comment (mikee47) [inline samples/Event_Jsvm/src/include/vm_functions.h]: remove

File: samples/Event_Jsvm/src/include/vm_functions.h [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+ */
+
+#pragma once
+
+#include <jerry-core/include/jerryscript.h>
+#include <WHashMap.h>
+#include <WString.h>
+
+#ifdef __cplusplus
+
```
Comment (mikee47) [inline samples/Event_Jsvm/src/include/vm_functions.h]: remove

File: samples/Event_Jsvm/src/vm_functions.cpp [hunk 10]
```diff
@@ -0,0 +1,110 @@
...
+	jerry_release_value(namePropertyValue);
+	jerry_release_value(nameProperty);
+
+	jerry_release_value(paramsPropertyName);
+	jerry_release_value(paramsPropertyValue);
+	jerry_release_value(paramsProperty);
+
+	jerry_release_value(eventObject);
+
+	//	bool is_error = jerry_value_is_error(res);
```
Comment (mikee47) [inline samples/Event_Jsvm/src/vm_functions.cpp]: remove

File: samples/Event_Jsvm/src/vm_functions.cpp [hunk 11]
```diff
@@ -0,0 +1,110 @@
...
+
+} // namespace
+
+jerry_value_t addEventListener(const jerry_value_t function_obj, const jerry_value_t this_val,
+							   const jerry_value_t args_p[], const jerry_length_t args_cnt)
+{
+	(void)function_obj; /* unused */
+	(void)this_val;		/* unused */
+
+	// First Parameter - Event Name
```
Comment (mikee47) [inline samples/Event_Jsvm/src/vm_functions.cpp:25]: Should check args_cnt == 2

**Review by mikee47 — CHANGES_REQUESTED** [review]
We should implement some C++ wrappers for handling values, etc. I've found `https://github.com/infosia/Daisy` and `https://gist.github.com/gocarlos/0c6783646074144b3af008b54a379d07` which consider at this issue.

Also some kind of wrapper mechanism which deals with parameter checking on function implementations. It might be possible to leverage C++ templates.

**Comment by slaff** [comment]
> We should implement some C++ wrappers for handling values, etc. I've found https://github.com/infosia/Daisy and https://gist.github.com/gocarlos/0c6783646074144b3af008b54a379d07 which consider at this issue. .. Also some kind of wrapper mechanism which deals with parameter checking on function implementations. It might be possible to leverage C++ templates.

I like the idea. The easier to use the better :)

**Pull request merged**